libcaf_core/caf/flow/op/mcast.hpp

```diff
@@ -192,6 +204,23 @@ class mcast : public hot<T> {
     }
   }
 
+  // -- implementation of ucast_sub_state_listener -----------------------------
+
+  void on_disposed(state_type* ptr) final {
+    super::ctx_->delay_fn([mc = strong_this(), sptr = state_ptr_type{ptr}] {
+      if (auto i = std::find(mc->states_.begin(), mc->states_.end(), sptr);
+          i != mc->states_.end()) {
+        // We don't care about preserving the order of elements in the vector.
+        // Hence, we can swap the element to the back and then pop it.
+        auto last = mc->states_.end() - 1;
+        if (i != last)
+          std::swap(*i, *last);
```

Comment:
I would rather not have an implementation detail leak to the client side. If we ever change the implementation or use a different smart pointer, this would break right away.

Reply:
Sure thing, let's keep the defensive approach.